pkg/cloudprovider/providers/gce/gce_loadbalancer_external_test.go

```diff
@@ -237,3 +243,238 @@ func TestDeleteAddressWithWrongTier(t *testing.T) {
 		})
 	}
 }
+
+const (
+	gceProjectId = "test-project"
+	gceRegion    = "test-region"
+	zoneName     = "zone1"
+	nodeName     = "test-node-1"
+	clusterName  = "Test Cluster Name"
+	clusterID    = "test-cluster-id"
+)
+
+var apiService = &v1.Service{
+	Spec: v1.ServiceSpec{
+		SessionAffinity: v1.ServiceAffinityClientIP,
+		Type:            v1.ServiceTypeClusterIP,
+		Ports:           []v1.ServicePort{{Protocol: v1.ProtocolTCP, Port: int32(123)}},
+	},
+}
+
+func fakeGCECloud() (*GCECloud, error) {
+	client, err := newOauthClient(nil)
+	if err != nil {
+		return nil, err
+	}
+
+	service, err := compute.New(client)
+	if err != nil {
+		return nil, err
+	}
+
+	fakeManager := newFakeManager(gceProjectId, gceRegion)
+
+	alphaFeatureGate, err := NewAlphaFeatureGate([]string{})
+	if err != nil {
+		return nil, err
+	}
+
+	cloud := cloud.NewMockGCE()
+	cloud.MockTargetPools.AddInstanceHook = mock.AddInstanceHook
+	zonesWithNodes := createNodeZones([]string{zoneName})
+
+	gce := GCECloud{
+		region:             gceRegion,
+		service:            service,
+		manager:            fakeManager,
+		managedZones:       []string{zoneName},
+		projectID:          gceProjectId,
+		AlphaFeatureGate:   alphaFeatureGate,
+		nodeZones:          zonesWithNodes,
+		nodeInformerSynced: func() bool { return true },
+		c:                  cloud,
+	}
+
+	return &gce, nil
+}
+
+func createExternalLoadBalancer(gce *GCECloud) (*v1.LoadBalancerStatus, error) {
+	err := gce.InsertInstance(
+		gceProjectId,
+		zoneName,
+		&compute.Instance{
+			Name: nodeName,
+			Tags: &compute.Tags{
+				Items: []string{nodeName},
+			},
+		},
+	)
+
+	if err != nil {
+		return nil, err
+	}
+
+	nodes := []*v1.Node{
+		{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: nodeName,
+				Labels: map[string]string{
+					kubeletapis.LabelHostname: nodeName,
+				},
+			},
+			Status: v1.NodeStatus{
+				NodeInfo: v1.NodeSystemInfo{
+					KubeProxyVersion: "v1.7.2",
+				},
+			},
+		},
+	}
+
+	status, err := gce.ensureExternalLoadBalancer(
+		clusterName,
+		clusterID,
+		apiService,
+		nil,
+		nodes,
+	)
+
+	return status, err
+}
+
+func TestEnsureExternalLoadBalancer(t *testing.T) {
+	gce, err := fakeGCECloud()
+	require.NoError(t, err)
+
+	status, err := createExternalLoadBalancer(gce)
+	assert.NoError(t, err)
+	assert.NotEmpty(t, status.Ingress)
+
+	lbName := cloudprovider.GetLoadBalancerName(apiService)
+
+	// Check that Firewall is created
+	firewall, err := gce.GetFirewall(MakeFirewallName(lbName))
```

Comment:
There are actually two firewalls created. One for the particular LB and one for allowing the health checker to hit kube-proxy. The latter firewall is shared among all LBs.

Reply:
Oh, found it - will add checks that the healthcheck firewall is created here + the delete lb test